### PR TITLE
Add deprecated v1 methods to BaseModel

### PR DESCRIPTION
**Pull request by dmontagu**
I'll note that I have not copied tests over from v1 for the newly-added methods. Let me know if you have opinions about if/how we do that.

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -342,72 +338,6 @@ def model_construct(cls: type[Model], _fields_set: set[str] | None = None, **val
...
-            value = getattr(self, name, Undefined)
-            if value is not Undefined:
-                if deep:
-                    value = deepcopy(value)
-                _object_setattr(m, name, value)
-
-        return m
-
-    # @typing_extensions.deprecated('This method is now deprecated; use `model_copy` instead')
-    def copy(
```
Comment (dmontagu) [inline pydantic/main.py:364]: I moved this and all other from-v1-deprecated methods to the bottom of the class in the same order they existed in v1.

I would also like to reorder the other (non-deprecated) methods of BaseModel, but will do that as a separate PR at an opportune time to reduce chances for merge conflicts etc.

File: pydantic/main.py [hunk 2]
```diff
@@ -810,6 +564,383 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+                proto=proto,
+                content_type=content_type,
+                encoding=encoding,
+                allow_pickle=allow_pickle,
+                json_loads=cls.model_config['json_loads'],
+            )
+        except (ValueError, TypeError, UnicodeDecodeError) as e:
+            # TODO: this used to raise ValidationError([ErrorWrapper(e, loc=ROOT_KEY)], cls)
+            #   Do we want to make it possible to instantiate from python, just for deprecated functionality?
+            raise e
```
Comment (dmontagu) [inline pydantic/main.py:643]: See TODO: comment above

File: pydantic/main.py [hunk 3]
```diff
@@ -810,6 +564,383 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+            'and use `model_validate` instead.',
+            DeprecationWarning,
+        )
+        if not cls.model_config['from_attributes']:
+            raise PydanticUserError('You must set the config attribute `from_attributes=True` to use from_orm')
+        return cls.model_validate(obj)
+
+    @classmethod
+    def construct(cls: type[Model], _fields_set: set[str] | None = None, **values: Any) -> Model:
+        warnings.warn('The `construct` method is deprecated; use `model_construct` instead.', DeprecationWarning)
```
Comment (dmontagu) [inline pydantic/main.py:682]: I have not copied over the v1 docstrings. Is that okay, or do we want to do that? I'm hoping the deprecation warning pointing the user at the other method is enough?

Comment (samuelcolvin) [reply pydantic/main.py:682]: Agreed.

File: pydantic/_internal/_deprecated/_parse.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+def load_str_bytes(
+    b: str | bytes,
+    *,
+    content_type: str = None,
+    encoding: str = 'utf8',
+    proto: Protocol = None,
+    allow_pickle: bool = False,
+    json_loads: Callable[[str], Any] = json.loads,
+) -> Any:
+    if proto is None and content_type:
```
Comment (samuelcolvin) [inline pydantic/_internal/_deprecated/_parse.py:24]: These should have warning?

File: pydantic/main.py [hunk 5]
```diff
@@ -810,6 +564,383 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+        *,
+        include: IncEx = None,
+        exclude: IncEx = None,
+        by_alias: bool = False,
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
+        exclude_none: bool = False,
+        # TODO: What do we do about the following arguments?
+        #   Do they need to go on model_config now, and get used by the serializer?
+        encoder: typing.Callable[[Any], Any] | None = None,
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/main.py:599]: I guess error if you use it. We need to point to docs on what to use instead.

File: pydantic/main.py [hunk 6]
```diff
@@ -810,6 +564,383 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+        from .json import pydantic_encoder
+
+        return cls.model_config['json_dumps'](
+            cls.model_json_schema(by_alias=by_alias, ref_template=ref_template),
+            default=pydantic_encoder,
+            **dumps_kwargs,
+        )
+
+    @classmethod
+    def validate(cls: type[Model], value: Any) -> Model:
```
Comment (samuelcolvin) [inline pydantic/main.py:747]: This was internal and wasn't documented, not sure we need it anymore.

File: pydantic/main.py [hunk 7]
```diff
@@ -810,6 +564,383 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+
+    @classmethod
+    def update_forward_refs(cls, **localns: Any) -> None:
+        warnings.warn(
+            'The `update_forward_refs` method is deprecated; use `model_rebuild` instead.', DeprecationWarning
+        )
+        cls.model_rebuild(force=True, types_namespace=localns)
+
+    # Deprecated: This private method is only used for `BaseModel.copy`, which is deprecated
+    def _iter(
```
Comment (samuelcolvin) [inline pydantic/main.py:759]: Can we move all this logic into the deprecated module as standalone functions?

Comment (dmontagu) [reply pydantic/main.py:759]: good idea, I'll do that.

File: pydantic/_internal/_deprecated/_parse.py
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/_internal/_deprecated/_parse.py:1]: good point, I'll just put it in a top-level `deprecated` module. Should it be `deprecated` or `_deprecated`?

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/_internal/_deprecated/_parse.py:1]: I guess just `deprecated`, it's not private.

File: pydantic/main.py (hunk 6 above)
Comment (samuelcolvin) [reply pydantic/main.py:747]: My I think we just raise an error, there's no obvious replacement for it.

File: pydantic/_internal/_deprecated/_parse.py
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/_internal/_deprecated/_parse.py:1]: I went with `deprecated` for now

File: pydantic/main.py (hunk 7 above)
Comment (dmontagu) [reply pydantic/main.py]: Moved these into deprecated/copy_internals.py, and retained the methods on the class with deprecation warnings. Let me know if you also want the methods themselves to have deprecation warnings. (If we do that, I think it may mean that calls to `Model.copy` trigger a bunch of warnings though)

File: pydantic/main.py (hunk 6 above)
Comment (dmontagu) [reply pydantic/main.py:749]: Looking at the definition from v1, it seems `model_validate` does the same thing for the most part: https://github.com/pydantic/pydantic/blob/11f7f9019016659e81c96481ffe5c42637489cf3/pydantic/main.py#L710-L719 

I realize there is the logic at the top related to what to do with subclass instances but that too seems handled by `model_validate` now with the recent changes you made in pydantic core. Is it really different enough to be worth removing instead of just adding a deprecation warning? I could add a note in the warning that the behavior may now be slightly different.

If you really think it's better to remove it, I'd suggest raising an error that says to replace it with model_validate, but I'm not sure what type of error to raise

Comment (samuelcolvin) [reply pydantic/main.py:749]: Ok, let's go with a warning.

**Comment by samuelcolvin** [comment]
conflicts and comments, please update.

File: pydantic/main.py [hunk 8]
```diff
@@ -810,6 +564,223 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+        cls: type[Model],
+        b: str | bytes,
+        *,
+        content_type: str = None,
+        encoding: str = 'utf8',
+        proto: _deprecated_parse.Protocol = None,
+        allow_pickle: bool = False,
+    ) -> Model:
+        warnings.warn(
+            'The `parse_raw` method is deprecated; load the data and use `model_validate` instead.', DeprecationWarning
```
Comment (samuelcolvin) [inline pydantic/main.py:629]: need to say

> use `model_validate_json` if your input is JSON, otherwise parse the data in your own code, and validate using `model_validate`

File: pydantic/main.py [hunk 9]
```diff
@@ -810,6 +564,223 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+        cls: type[Model],
+        path: str | Path,
+        *,
+        content_type: str = None,
+        encoding: str = 'utf8',
+        proto: _deprecated_parse.Protocol = None,
+        allow_pickle: bool = False,
+    ) -> Model:
+        warnings.warn(
+            'The `parse_file` method is deprecated; load the data and use `model_validate` instead.', DeprecationWarning
```
Comment (samuelcolvin) [inline pydantic/main.py:657]:
```suggestion
            'The `parse_file` method is deprecated; load the data and use `model_validate` instead.', DeprecationWarning
```
as above, recommend `model_validate_json`.

File: pydantic/main.py [hunk 10]
```diff
@@ -810,6 +564,223 @@ def model_parametrized_name(cls, params: tuple[type[Any], ...]) -> str:
...
+            encoding=encoding,
+            allow_pickle=allow_pickle,
+            json_loads=cls.model_config['json_loads'],
+        )
+        return cls.parse_obj(obj)
+
+    @classmethod
+    def from_orm(cls: type[Model], obj: Any) -> Model:
+        warnings.warn(
+            'The `from_orm` method is deprecated; set model_config["from_attributes"]=True '
```
Comment (samuelcolvin) [inline pydantic/main.py:672]:
```suggestion
            'The `from_orm` method is deprecated; set `model_config["from_attributes"]=True` '
```

Comment (samuelcolvin) [reply pydantic/main.py:672]: I guess we can remove the note about `model_config["from_attributes"]=True` if they already have it set and we're not raising an error.

File: pydantic/deprecated/copy_internals.py
Comment (samuelcolvin) [inline pydantic/deprecated/copy_internals.py:1]: Might be worth removing `pydantic/deprecated` from coverage completely.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM, either now or soon I think it's worth adding at least one test for each function to a new `tests/test_depreciated.py` module. Might move `test_orm_mode.py` tests into there too.

File: pydantic/main.py (hunk 2 above)
Comment (samuelcolvin) [reply pydantic/main.py:643]: https://github.com/pydantic/pydantic-core/pull/510

File: pydantic/main.py (hunk 10 above)
Comment (samuelcolvin) [reply pydantic/main.py:672]: on reflection, I think this is fine.
